magpie/services.py

```diff
@@ -72,85 +65,315 @@ def __init__(self, service, request):
-                for outcome, perm_user, perm_name in permission_to_pyramid_acls(permissions):
-                    self.acl.append((outcome, perm_user, perm_name,))
+        .. seealso::
+            - :meth:`ServiceInterface.resource_requested`
+        """
+        if not permissions:
+            permissions = self.allowed_permissions(resource)
+        requested_perms = set(permissions)  # type: Set[Permission]
+        effective_perms = dict()            # type: Dict[Permission, PermissionSet]
+
+        # immediately return all permissions if user is an admin
+        db_session = self.request.db
+        admin_group = get_constant("MAGPIE_ADMIN_GROUP", self.request)
+        admin_group = GroupService.by_group_name(admin_group, db_session=db_session)
+        if admin_group in user.groups:  # noqa
+            return [PermissionSet(perm, Access.ALLOW, Scope.MATCH, PermissionType.EFFECTIVE) for perm in permissions]
+
+        # current and parent resource(s) recursive-scope
+        match = allow_match
+        while resource is not None:  # bottom-up until service is reached
+
+            # include both permissions set in database as well as defined directly on resource
+            cur_res_perms = ResourceService.perms_for_user(resource, user, db_session=db_session)
+            cur_res_perms.extend(permission_to_pyramid_acls(resource.__acl__))
+
+            for perm_name in requested_perms:
+                for perm_tup in cur_res_perms:
+                    perm_set = PermissionSet(perm_tup)
+
+                    # if user is owner (directly or via groups), all permissions are set,
+                    # but continue processing this resource until end in case user explicit deny reverts it
+                    if perm_tup.perm_name == ALL_PERMISSIONS:
+                        for perm in requested_perms:
+                            all_perm = PermissionSet(perm, perm_set.access, perm.scope, perm.type)
+                            if perm_set.access == Access.DENY:
+                                effective_perms[perm] = all_perm
+                            else:
+                                effective_perms.setdefault(perm, all_perm)
+                    # skip if the current permission must not be processed (at all or for the moment until next iter)
+                    elif perm_set.name not in requested_perms or perm_set.name != perm_name:
+                        continue
+                    # only first resource can use match, parents are recursive-only
+                    if not match and perm_set.scope == Scope.MATCH:
+                        continue
+
+                    # user direct permissions have priority over inherited ones from groups
+                    if perm_set.type == PermissionType.DIRECT:
+                        perm = effective_perms.get(perm_name)
+                        # explicit user deny overrides user allow if any already found
```

Comment:
The if seems to also include the case where we have an explicit allow that also override inherited permissions (whatever the type). Should also be in the comment. I know that previous comment indicate that. But given the complexity of the algo, I think that the comment should match exactly what the if does.